### DISCUSSION TITLE
API, python datamodel structure and ReqIF standard implementation questions

### DISCUSSION THREAD

**Discussion by mikelvin** [strictdoc-project/reqif#135] [General]
Hello,

I am a student currently working on understanding and using the ReqIF standard. Firstly, I must apologize because this is my first interaction with a developer community that is working on an open source project, as is this reqif library. I'm also an English student and beginner programmer who has almost finished a computer science degree, so it is possible that many things said in this post could be wrong. If that's the case, do not doubt correcting me, I am willing to learn from you all.

My objective is to generate, using some external data, a ReqIF standard and compliant file. To accomplish this, I have been looking for different libraries and a few weeks ago I came across this one. 

I have been reading the documentation available on your site and I also have been browsing the existing code to understand the data model and the reqif lib API. In this process some questions came into my mind and I would be so grateful to you if you could solve them for me. 

- I tried to use the existing reqif lib API to create a brand new reqif file but the code I was creating was becoming so complicated. I thought that the reqif library user (me) had to manage some of the responsibilities that the reqif API should have been taking. Due to the lack of knowledge I have on the main objectives of this library, I don't know if a more elaborated API could be a future target objective. In the case of an existing interest in updating the API, I would be interested in the development process of it.

- Why does not the python reqif lib datamodel follow the data model specified in the ReqIF standard specification? I mean, I know that there are multiple ways to implement a standard but, if the standard is observed, there is a straightforward way in which a python equivalent data model could be implemented. In the current reqif python library implementation there are some relations, attributes, class names and class hierarchies (extensions or inheritance) that do not match with the standard. Would it be a good decision to refactor some class attributes or class names and also add inheritance mechanisms? I think that this could allow us to define more accurately the typing of attributes and also would ease the extension of API functionality creation and data model understanding.

- This next question is closely related to the previous one but, why does the reqif library manage references using just identifiers in string form instead of using the object references themselves? I think that it could be due to the architecture design but, I want to understand the advantages and disadvantages taken into account to make that design choice. 

Thank you all in advance!

Mikel :)

**Comment by stanislaw** [comment]
Hi @mikelvin ,

> I am a student currently working on understanding and using the ReqIF standard. Firstly, I must apologize because this is my first interaction with a developer community that is working on an open source project, as is this reqif library. I'm also an English student and beginner programmer who has almost finished a computer science degree, so it is possible that many things said in this post could be wrong. If that's the case, do not doubt correcting me, I am willing to learn from you all.

Thank you for your interest in this library, and welcome to the open source world! You are asking good questions, and I will try to explain to you how the project came to be where it is now. I will reorder your questions, so that my answers appear in a more general-to-specific order.

> Why does not the python reqif lib datamodel follow the data model specified in the ReqIF standard specification? I mean, I know that there are multiple ways to implement a standard but, if the standard is observed, there is a straightforward way in which a python equivalent data model could be implemented. In the current reqif python library implementation there are some relations, attributes, class names and class hierarchies (extensions or inheritance) that do not match with the standard. Would it be a good decision to refactor some class attributes or class names and also add inheritance mechanisms? I think that this could allow us to define more accurately the typing of attributes and also would ease the extension of API functionality creation and data model understanding.

Some background info on this project: maintaining this library is not my full-time job. I work on this and other projects in my spare time (think 10 hours per week on average). Furthermore, this reqif library is a child project of my other project StrictDoc where I focus most of my time. Given this setup, the library already got much further than many other Python ReqIF scripts found on GitHub, and there is a good potential for pushing it closer to the ReqIF standard.

With these time constraints, I created this library to be able to convert basic ReqIF files to and from my tool's SDoc format and I started working with only basic ReqIF examples randomly found on the internet. I looked into the ReqIF standard's spec and decided to not! do a top-down implementation due to the lack of time. Instead, I focused on the bare minimum needed to get the roundtrip ReqIF file -> ReqIF python objects -> ReqIF file to work on the basic examples and then more complicated ones.

The existing implementation is a bottom-up one with the advantage that it is pretty well tested. The existing object fields are quite close to those of the ReqIF standard, so the library is not very far from being ReqIF compliant.

Specifically to your `does not the python reqif lib datamodel follow the data model specified` part, could you summarize which discrepancies did you notice and what would be your proposed fixes to these problems? The implementation in its current state works well for my needs but I am happy to learn about everything I might have missed.

> My objective is to generate, using some external data, a ReqIF standard and compliant file. To accomplish this, I have been looking for different libraries and a few weeks ago I came across this one.

What do you mean by `generate a ReqIF standard`? Or am I misreading, and you mean that your file must be compliant to the ReqIF standard (standard and compliant are kind of synonyms, aren't they)?

> I have been reading the documentation available on your site and I also have been browsing the existing code to understand the data model and the reqif lib API. In this process some questions came into my mind and I would be so grateful to you if you could solve them for me.

> I tried to use the existing reqif lib API to create a brand new reqif file but the code I was creating was becoming so complicated. 

Originally, this library has not been designed with the goal of creating new ReqIF files. The main use case was to parse the existing ReqIF files or convert to ReqIF files from some other formats. My main use case is SDoc format, see [example](https://github.com/strictdoc-project/strictdoc/blob/main/docs/strictdoc_01_user_guide.sdoc) of the format and the [examples of the from- and to-ReqIF converters](https://github.com/strictdoc-project/strictdoc/tree/main/strictdoc/backend/reqif/p01_sdoc). In my case, I needed a low-level access to all the ReqIF objects details as I was learning through the object structure assumed by ReqIF.

The feature of creating new ReqIF files with a more high-level and nicer API can be implemented, and we could certainly work on creating a good API for that.

> I thought that the reqif library user (me) had to manage some of the responsibilities that the reqif API should have been taking. Due to the lack of knowledge I have on the main objectives of this library, I don't know if a more elaborated API could be a future target objective. In the case of an existing interest in updating the API, I would be interested in the development process of it.

I would be happy to improve the existing or add the new API, and this is where you contribution would be appreciated. I would love to see a specific proposal and better several independent proposals if you have multiple ideas on how to improve the API. Reviewing the proposals independently would help us to focus on them one at a time.

> This next question is closely related to the previous one but, why does the reqif library manage references using just identifiers in string form instead of using the object references themselves? I think that it could be due to the architecture design but, I want to understand the advantages and disadvantages taken into account to make that design choice.

The answer is pretty much the same: what you see in the existing library is what works well for me and several other users who reported to be happy about the current state of the API.

Could you show a code fragment that is not working well for you, and how would you like to have the API look like?

Thanks for your questions, and I am looking forward to learning about your ideas.

**Comment by stanislaw** [comment]
Hey @mikelvin, you asked very good questions, and I am still curious to know if you are going to resume this thread in one way or the other. In case you decided to not focus your efforts on reqif, please let me know as well, so that I know what to expect. Thank you.

**Comment by mikelvin** [comment]
Hello @stanislaw,

Sorry for not responding. I have been and am currently very busy working on some important project and I have not had the necessary time to construct the adequate answer that I would have wanted to give you.

I would really like to help on the development of the ReqIf python library but it would consume time that I don't really have at this moment due to the other tasks that I must address. 

So, for now I will be out of this interesting project until I manage to have some free time I would like to spend here.

Thank you for answering my questions, for your interest and for your involvement in every aspect of this project.

**Comment by stanislaw** [comment]
Hi @mikelvin ,

Thanks for coming back. In fact, you don't have to go hard on this topic and do any work yourself. I was wondering if you could simply sketch out a short note on what would be the top three improvements that you would like to see implemented in the library.

I develop this library in iterations. If you would provide your input, I could simply grab and implement some of your ideas at one of my next passes to it.

Thank you.

**Comment by mcrist5564** [comment]
I am also new to the ReqIF standard from a parsing perspective. From my very short experience with this library, I have three simple improvements that I would ask for:

1. Align the names of the python in-memory objects to the reqif node names - The first example of this is the `<SPEC-HIERARCHY><SPEC-OBJECT-REF>`...in the library, this is spec_object instead of spec_object_ref. I may be reading too much into the request from the OP, but this caught me off guard as I'm getting started.
2. Make it very obvious how to retrieve the requirement text - I'm still not sure how to do this, and will continue to read through the examples and unit tests. It's not clear how to navigate from a SPEC-OBJECT-REF to the SPEC-OBJECT with the same identifier to retrieve the LONG-NAME or THE-VALUE. Even once I can retrieve THE-VALUE, most of it is in XHTML, whereas I want plain text.
3. Get the object browser to work properly - I'm not sure how to ask this...I'm using pycharm, and usually try to use the word completion features. When I type the name of an object and hit . the list of completion options comes up. This is how I learn what I can do with that object, and can usually figure out from the list that comes up what operations I'm interested in. Specification types of objects have a good list:
![image](https://github.com/strictdoc-project/reqif/assets/103584163/5725aad8-e23e-474f-9010-5f35724ca10c)
but hierarchy objects don't:
![image](https://github.com/strictdoc-project/reqif/assets/103584163/c91df902-bbd6-4bd8-a2e8-1c5d762414c3)

I hope that makes sense.

-edit: cleanup code tags

**Reply by stanislaw** [reply]
Hey @mcrist5564 ,

For your first two points, I have created two issues: https://github.com/strictdoc-project/reqif/issues/142 and https://github.com/strictdoc-project/reqif/issues/143. Both could be addressed, and I will get to them at some point in time.

The third point seems to be the easiest to address: https://github.com/strictdoc-project/reqif/issues/144. The function is simply missing a type hint as follows:

Now:

```
    def iterate_specification_hierarchy(self, specification) -> Generator:
```

Should be:

```
    def iterate_specification_hierarchy(self, specification) -> Generator[ReqIFSpecification]:
```

I will check some other function signatures to ensure that the type hints are there. Thanks for reporting all this 👍
